pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy

```diff
@@ -690,15 +690,12 @@ class ModelValidatorImpl implements ModelValidator {
             stepsStagesParallelCount += 1
         }
 
-        if (isWithinParallel && (stage.branches.size() > 1 || stage.parallel != null)) {
+        if (isWithinParallel && (stage.branches.size() > 1 || stage.parallel != null || stage.matrix != null)) {
```

Comment:
At this point, it may be worth adding a method to `ModelASTStage` that does the `(stage.branches.size() > 1 || ... )` and just calling that here.

Reply:
Yeah, some clearer behavior....

Reply:
I tried it and it doesn't make it clearer... I'm going to leave it for now.